Add math styling module

mkorje opened the pull request:
This PR adds functions and the related infrastructure needed to take a character/string and a style in math (e.g. blackboard-bold, monospace), and return it in styled form. It is implemented under the feature `styling`, which I've added as a default.

Moving this to Codex was brought up briefly on Discord, and whilst it doesn't directly fall under "naming Unicode symbols", I think it still fits quite nicely in this crate. I may have jumped the gun a little on this, but it wasn't much work to do. Hopefully this PR starts some more thorough discussion of this idea. I also think that moving this outside of the main typst repo will be an improvement and make it more maintainable.

MDLC01 commented:
A bit more context: https://github.com/typst/codex/issues/5#issuecomment-2488669690.

MDLC01 commented on src/styling.rs:
```diff
@@ -0,0 +1,613 @@
+
+/// Functions which map a [`char`] to its specified styled form.
+///
+/// Sourced from:
+/// - [Mathematical Alphanumeric Symbols]
+/// - [Arabic Mathematical Alphabetic Symbols]
+/// - <https://www.w3.org/TR/mathml-core/#new-text-transform-mappings>
```
This should probably link the Unicode Core Specification as well. For example, <https://www.unicode.org/versions/Unicode16.0.0/core-spec/chapter-22/#G52612>.

MDLC01 commented on src/styling.rs:
```diff
@@ -0,0 +1,613 @@
+    /// upon the font.
+    pub fn to_serif_bold(c: char) -> char {
+        let delta = match c {
+            'A'..='Z' => 0x1D3BF,
+            'a'..='z' => 0x1D3B9,
+            'Α'..='Ρ' => 0x1D317,
+            'ϴ' => 0x1D2C5,
+            'Σ'..='Ω' => 0x1D317,
+            '∇' => 0x1B4BA,
+            'α'..='ω' => 0x1D311,
+            '∂' => 0x1B4D9,
+            'ϵ' => 0x1D2E7,
+            'ϑ' => 0x1D30C,
+            'ϰ' => 0x1D2EE,
+            'ϕ' => 0x1D30A,
+            'ϱ' => 0x1D2EF,
+            'ϖ' => 0x1D30B,
+            'Ϝ' | 'ϝ' => 0x1D3EE,
+            '0'..='9' => 0x1D79E,
```
I think it would be better to somehow use characters (e.g., `'𝐀'`) instead of codepoints on the right hand sides here.

mkorje replied:
At the moment the right hand side is just a delta to add to the original codepoint. I can't quite remember why exactly I did it this way instead of just mapping directly to the new codepoint, in which case we could use characters instead.

T0mstone replied:
The delta allows the pattern to be `'A'..='Z'` instead of 26 individual ones.

mkorje replied:
Oh of course, for some reason I thought `'A'..='Z' => '𝐀'..='𝐙'` would work :sweat_smile:.

MDLC01 commented on src/styling.rs:
```diff
@@ -0,0 +1,613 @@
+    ///
+    /// This function returns a variation sequence using the variation selector
+    /// `U+FE00`, as specified in the [`StandardizedVariants.txt`] file from
+    /// the Unicode Character Database. Note that only the capital latin
+    /// letters are standardized variation sequences, but this function also
+    /// maps the small latin letters.
```
```suggestion
    /// the Unicode Character Database. Note that only the capital Latin
    /// letters are standardized variation sequences, but this function also
    /// maps the small Latin letters.
```

And similar in three other places.

MDLC01 commented:
Regardless of my above comments, @laurmaedje will have to approve moving those functions before we merge this PR.

mkorje commented:
> Regarding the implementation, as well as having many `to_{style}` functions, there could also be a `MathStyle::apply` method to apply a style to a `char`.

The issue I had with something like this is that some of the styles will output multiple characters.

I also considered implementing the `MathStyling` trait for `EcoString`, as `EcoString` is used in the Typst repo. Though I wasn't sure about adding that as a dependency; maybe instead it could be under a feature flag?

MDLC01 commented:
> > Regarding the implementation, as well as having many `to_{style}` functions, there could also be a `MathStyle::apply` method to apply a style to a `char`.
> 
> The issue I had with something like this is that some of the styles will output multiple characters.

Can you not return `[char; 2]` like the `to_style` function? Essentially, making `to_style` a method on `MathStyle`.

> I also considered implementing the `MathStyling` trait for `EcoString`, as `EcoString` is used in the Typst repo. Though I wasn't sure about adding that as a dependency; maybe instead it could be under a feature flag?

Instead, maybe `to_styled` could be made generic over its return type `T: Default + AddAssign<&str>`. Or maybe I'm just overthinking, and we can just add the dependency behind a feature flag.

MDLC01 commented:
Also, you can implement `MathStyling` for `str` instead, that way anything that derefs to `str` (such as `String` or `EcoString`) can benefit from `to_styled`.

laurmaedje commented:
Looks like there's quite a bit more going on here than in typst-layout/src/math/text.rs, e.g. the looped and chancery stuff. Would that somehow be exposed in Typst or is it for completeness?

laurmaedje commented:
Regarding EcoString: I wouldn't add a dependency on it here. Doesn't seem worth it to me.

laurmaedje commented on src/styling.rs:
```diff
@@ -0,0 +1,616 @@
+/*!
+Style mathematical symbols in Unicode.
+*/
```
```suggestion
//! Style mathematical symbols in Unicode.
```

laurmaedje commented on src/styling.rs:
````diff
@@ -0,0 +1,616 @@
+/// assert_eq!(['ⅈ', '\0'], to_style('i', MathStyle::DoubleStruckItalic));
+/// assert_eq!(['𝓴', '\u{fe01}'], to_style('k', MathStyle::RoundhandBold));
+/// ```
+pub fn to_style(c: char, style: MathStyle) -> [char; 2] {
````
The `[char: 2]` approach where `\0` has special meaning seems a bit easy to misuse to me. Maybe it could return a custom iterator type instead? Then to string impl would be trivial to do yourself (`s.chars().flat_map(|c| to_style(c, style)).collect()`. I would probably even remove the trait then.

mkorje replied:
I agree an iterator is much nicer. I've basically mimicked [`to_uppercase`](https://doc.rust-lang.org/std/primitive.char.html#method.to_uppercase) and [`ToUppercase`](https://doc.rust-lang.org/std/char/struct.ToUppercase.html) from `std`. I also got rid of the trait.

mkorje commented:
> Looks like there's quite a bit more going on here than in typst-layout/src/math/text.rs, e.g. the looped and chancery stuff. Would that somehow be exposed in Typst or is it for completeness?

Yes, I plan to expose these in Typst as well. At this point in time it is just for completeness though.

For the Arabic ones I plan to add to [Variants](https://typst.app/docs/reference/math/variants) the functions `inital`, `tailed`, `looped`, and `stretched`, as well as `isolated` as an alias for `serif`. Though I should note I have no idea if these should even really have their names in English, or if shortened names like `tail` or `loop` would work.

As for the chancery/roundhand stuff, we'd add `scr` to join `cal` in Typst. But there is still some work to do before it can be added (cf. [typst/typst 5853](https://github.com/typst/typst/issues/5853)).

mkorje commented:
I've changed the `to_style` function to return a custom iterator type and have removed the `MathStyling` trait for now (see https://github.com/typst/codex/pull/20#discussion_r1966739628)

> Instead, maybe `to_styled` could be made generic over its return type `T: Default + AddAssign<&str>`. Or maybe I'm just overthinking, and we can just add the dependency behind a feature flag.

> Also, you can implement `MathStyling` for `str` instead, that way anything that derefs to `str` (such as `String` or `EcoString`) can benefit from `to_styled`.

Since I've removed the trait for now, this no longer applies.

> Regarding the implementation, as well as having many `to_{style}` functions, there could also be a `MathStyle::apply` method to apply a style to a `char`.

I think in my previous reply to this I misunderstood what you were saying, sorry. I've done that now.

Also, I was wondering whether to make the `to_style` function a method on `char` instead?

I also need to double check Arabic, as for the normal style the math symbols have separate codepoints (unlike latin where the upright serif for math is the same as the standard codepoints), I think.

laurmaedje commented:
I agree that we shouldn't have `[char; 2]` in the public API. I don't think we need a public `MathStyle::apply`. I think we can inline its implementation into `to_style`.

mkorje commented:
When you say not having `[char; 2]` in the public API, does that include the functions in `mappings`?

mkorje commented:
Ok I've redone some stuff and more thoroughly documented/commented everything, which makes the code all much clearer I think. I also double checked the deltas and cross referenced with MathML Core. I think this ready from my end to review.

> I hadn't seen that those are public. Feels slightly duplicate to me with the enum. It's nice to have have in the type that they return just one char. I guess it depends a bit on how you plan to use them in Typst.

I just decided to make those functions private in the end.

mkorje commented:
Oh there is one thing left: we should maybe (?) switch alef, bet, dalet, gimel to map to their mathematical glyphs instead of the Hebrew letters.

MDLC01 commented on src/styling.rs:
```diff
@@ -0,0 +1,788 @@
+use std::fmt::{self, Write};
+use std::iter::FusedIterator;
+
+/// The version of [Unicode](https://www.unicode.org/) that this version of the
```
```suggestion
/// The version of [Unicode](https://www.unicode.org/versions/Unicode16.0.0/core-spec/) that this version of the
```

This may be a more useful link.

mkorje replied:
Though that would then need updating with each version change as well. Not sure if this is really much of a point though.

MDLC01 replied:
Well, the tuple below will have to be updated as well anyway. But alternatively you can replace `Unicode16.0.0` with `latest`.

mkorje replied:
That could also cause issues as `latest` might not match the version here :sweat_smile:

MDLC01 commented on src/styling.rs:
```diff
@@ -0,0 +1,788 @@
+/// Notation in mathematics uses a basic set of characters which can be styled.
+/// The following groupings are used in the documentation:
+/// - digits: The basic Latin digits 0–9 (U+0030..U+0039).
+/// - latin: The basic uppercase and lowercase Latin letters, a–z
```
```suggestion
/// - Latin: The basic uppercase and lowercase Latin letters, a–z
```

"Latin" should be capitalized, same for "Greek" and "Arabic" below.

MDLC01 commented on src/styling.rs:
```diff
@@ -0,0 +1,788 @@
+    ///
+    /// Supported characters: latin, greek.
+    BoldItalic,
+    /// Script style.
```
May be chancery or roundhand depending on the font.

mkorje replied:
Good catch, will add that.

MDLC01 commented on src/styling.rs:
```diff
@@ -0,0 +1,788 @@
+    /// Sans-serif style.
+    ///
+    /// Supported characters: digits, latin.
+    Sansserif,
```
```suggestion
    SansSerif,
```

Not very important, but writing "sans serif" as one word feels weird to me. Another alternative is simply `Sans`.

T0mstone replied:
`Sans` is very familiar since it's present in tons of font names (e.g. Comic Sans), so I'd slightly prefer that, especially given the length of `SansBoldItalic` vs `SansSerifBoldItalic`.

mkorje replied:
IIRC my thought was that hyphens shouldn't be camel case otherwise it reads like two separate words "sans serif" instead of "sans-serif".

As with double-struck, despite being more verbose I think it'd be good to keep with the full name "sans-serif" used in Unicode.

laurmaedje replied:
I think `SansSerif` is definitely more readable than `Sansserif`

T0mstone replied:
> IIRC my thought was that hyphens shouldn't be camel case otherwise it reads like two separate words "sans serif" instead of "sans-serif".

It *is* two separate words. Sans means "without", so it's just saying "without serifs".

mkorje replied:
I think I meant that as two separate words they could then each read like a style modifier (e.g. Script and BoldScript): SansSerifBold is Sans-serif Bold not Sans (Serif Bold). Anyways I just made the change to SansSerif, and also to DoubleStruck.

T0mstone commented on src/styling.rs:
```diff
@@ -0,0 +1,788 @@
+    /// Supported characters: digits, latin, arabic excluding ا (U+0627),
+    /// ك (U+0643), ه (U+0647), and the extras ∑ (U+2211), Γ (U+0393), Π
+    /// (U+03A0), γ (U+03B3), π (U+03C0).
+    Doublestruck,
```
Same as with `Sansserif` here.
Either
```suggestion
 DoubleStruck,
```
or I'd also propose just using
```suggestion
 Double,
```

MDLC01 replied:
Simply "double" instead of "double struck" isn't nearly as common as "sans" for "sans serif", so I think we should keep "double struck" here.

mkorje replied:
The name used is double-struck in Unicode, so I figure its best to stick to that.

Enivex commented:
What is the status of this exactly?

mkorje commented:
I need to open a PR on the Typst side. I think I almost finished it a while ago, I'll finish it soon...

mkorje commented:
Ok, this should all be done for the codex side of things. See https://github.com/typst/typst/pull/6309 for the accompanying Typst PR.

The last commit moves the `MathVariant` enum from Typst here and adds some logic with the `resolve_style` function to decide on the style to use given options for the variant, bold, and italic (and improves a bit the user experience in Typst - more in the PR linked above). You'll notice that the Arabic variants and roundhand/chancery are missing from the `MathVariant` enum here. I'll only add them there once they're ready to be used in Typst, which isn't the case at the moment (and I didn't want that to block this PR). I guess when that does get updated it'd maybe warrant a minor version bump?

laurmaedje commented on src/styling.rs:
```diff
@@ -0,0 +1,867 @@
+            // Letterlike symbols (U+2100..U+2134)
+            'g' => 0x20A3,
+            'H' => 0x20C3,
+            'I' => 0x20C7,
```
are these deltas are computed by hand? I only now noticed that these numbers aren't codepoint numbers as in the previous code but rather deltas.

mkorje replied:
I think it's just one of the Arabic styles that was completely new and done by hand

laurmaedje replied:
Ah, the spec also defines it in terms of deltas?

mkorje replied:
It's technically not in the spec, but in the non-normative [C. Mathematical Alphanumeric Symbols](https://www.w3.org/TR/mathml-core/#new-text-transform-mappings). I have no idea where they have all gone, suddenly there's only italic there now. Anyways, it had both the original codepoint, new one, and the delta.

Delta just ends up being a bit nicer as there tend to be chunks that all have the same delta.

laurmaedje commented on src/styling.rs:
```diff
@@ -0,0 +1,867 @@
+    DoubleStruck,
+}
+
+pub fn resolve_style(
```
I think this might be nicer as a method `MathStyle::select`. A bit of docs would also be great, especially for the `None` cases and on the fallback behaviour.

mkorje replied:
Can do.
Though I was a bit unsure here with docs and all since it's basically going to just follow Typst. And so at the moment the new enum will be missing some of the styles initially.

mkorje replied:
The MathVariant one you left the other comment on.

laurmaedje replied:
I guess this piece of the code could also stay in Typst.

mkorje replied:
Though I'd like (and think it'd be nice) to not have any character matching code stuff on the Typst side

I'll leave out the Arabic styles from this enum for now (will add them in a later PR) and add the `non_exhaustive` attribute.

laurmaedje replied:
Since we've reframed the goal of codex to be more generally anything Typst i18n and Unicode related, I'm more fine with having things here be not totally generalized.

laurmaedje commented on src/styling.rs:
```diff
@@ -0,0 +1,867 @@
+    ToStyle::new(styled)
+}
+
+#[derive(Debug, PartialEq, Eq, Clone, Copy, Hash)]
```
Could you add one sentence of docs here as well, just so that all public items are documented?

mkorje commented:
I'm blocked by https://github.com/typst/typst/pull/6441 now on the Typst side as https://github.com/typst/codex/pull/86 is causing errors... I'm too tired to try sort this out tonight and I am like 75% confident this should be working on the Typst :laughing:

mkorje commented:
I should also remove Script from the MathVariants enum as Roundhand and Chancery are already on it.

MDLC01 commented:
Except script is different because it doesn't specify a variation selector, right? If so, we can probably keep it (even if we don't use it in Typst).

mkorje commented:
Yes, but if the font doesn't support it it just gets ignored.

I thought to remove it because we'd then not be using it in Typst, though I guess there's no harm in keeping it.

laurmaedje commented on src/styling.rs:
````diff
@@ -0,0 +1,942 @@
+    /// ```
+    pub fn select(
+        c: char,
+        variant: Option<MathVariant>,
````
Why is this an `Option` if `Plain` and `None` behave equivalently?
